runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java

```diff
@@ -12,86 +14,181 @@
-        reloadTrustStoreWrapper();
+        if (watcher.isCertExpiryCheckEnabled()) {
+            for (X509Certificate cert : chain) {
+                cert.checkValidity();
+            }
+        } else {
+            logCertExpiryCheck();
+        }
+
+        X509TrustManager trustManager = getTrustManager();
         trustManager.checkServerTrusted(chain, authType);
     }
 
+    private void logCertExpiryCheck() {
+        log.info(
+                "Certificate expiry check has been disabled with: {}",
+                watcher.trustStoreConfig.getDisableCertExpiryCheckFile()
+        );
+    }
+
     @Override
     public X509Certificate[] getAcceptedIssuers() {
         return new X509Certificate[0];
     }
 
-    /**
-     * Just a wrapper due to IDE pointing out duplicate code.
-     *
-     * @throws CertificateException Wrapper for any exception from reloading the trust store.
-     */
-    private void reloadTrustStoreWrapper() throws CertificateException {
+    private X509TrustManager getTrustManager() throws CertificateException {
+        X509TrustManager trustManager;
         try {
-            reloadTrustStore();
-        } catch (SSLException e) {
-            String message = "Unable to reload trust store " + trustStoreConfig.getTrustStoreFile() + ".";
-            throw new CertificateException(message, e);
+            trustManager = watcher.getTrustManager();
+        } catch (Exception e) {
+            throw new CertificateException(e);
         }
+        return trustManager;
     }
 
-    /**
-     * Reload the trust manager.
-     *
-     * @throws SSLException Thrown when there's an issue with loading the trust store.
-     */
-    private void reloadTrustStore() throws SSLException {
-        KeyStore trustStore = TlsUtils.openCertStore(trustStoreConfig);
+    public static class TrustStoreWatcher {
+        @NonNull
+        private final TrustStoreConfig trustStoreConfig;
 
-        TrustManagerFactory tmf;
-        try {
-            tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
-            tmf.init(trustStore);
-        } catch (NoSuchAlgorithmException e) {
-            String errorMessage = "No support for TrustManagerFactory default algorithm "
-                    + TrustManagerFactory.getDefaultAlgorithm() + ".";
-            throw new SSLException(errorMessage, e);
-        } catch (KeyStoreException e) {
-            String errorMessage = "Unable to load trust store " + trustStoreConfig.getTrustStoreFile() + ".";
-            throw new SSLException(errorMessage, e);
+        private final ExecutorService executor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("TrustStore-watcher-%d")
+                .build());
+
+        private volatile CompletableFuture<TrustManagerContext> trustManagerAsync;
```

Comment:
I didn't support the revert of the initial implementation because the fix was trivial and it should have been fixed instead of reverted, but now that it is reverted, maybe its a good time to discuss what this patch tries to achieve, is it performance? and does it really matter if this is just an initialization step and doesn't occur frequently. 


Furthermore, the bug that caused the revert was a deadlock and not starvation, but the git commit message says this patch fixes a starvation issue, are you using the two interchangeably or is there an additional starvation problem that wasn't surfaced with the deadlock bug?

Reply:
I think the handshake can take more than the 500 ms (i.e., default value) irrespective of this patch therefore that issue should be addressed by tuning the timeout value.

From what I understand, this patch optimizes the specific cases where two or more connections are initialized within a 3 second window, because the trust manager is loaded once every 3 seconds, then connections within that time frame can share the same store without needing to load it from the file system. 

I think this patch isn't necessary for the following reasons:
1. The client connections are long running and few
2. The optimization uses async APIs for what is inherently blocking logic
3. The file system latency isn't really decoupled from the handshake because even if the value is cached the file system is still read via TrustStoreConfig::isCertExpiryCheckEnabled()

In my opinion the complexity/optimization trade-off isn't worth it because the optimize scenario is infrequent and unlikely to happen. For example, look at this method which just reads a file `TlsUtils::openCertStore` , it's really convoluted because it introduces threading/coordination of what is simply sequential blocking code 

```java

public static CompletableFuture<KeyStore> openCertStore(CertStoreConfig certStoreCfg) {
        return CompletableFuture
                //validation step
                .runAsync(certStoreCfg::checkCertStoreExists, EXECUTOR)
                .thenCompose(empty -> getCertStorePassword(certStoreCfg.getPasswordFile()))
                .thenApplyAsync(certStorePassword -> {
                    KeyStore keyStore = getKeyStoreInstance();

                    Path certStore = certStoreCfg.getCertStore();
                    FileInputStream inputStream = null;
                    try {
                        inputStream = new FileInputStream(certStore.toFile());
                        keyStore.load(inputStream, certStorePassword.toCharArray());
                        return keyStore;
                    } catch (IOException e) {
                        String errorMessage = String.format("Unable to read key store file %s.", certStore);
                        throw new IllegalStateException(errorMessage, e);
                    } catch (CertificateException e) {
                        String errorMessage = String.format("Unable to read certificates in key store file %s.", certStore);
                        throw new IllegalStateException(errorMessage, e);
                    } catch (NoSuchAlgorithmException e) {
                        String errorMessage = String.format("No support for algorithm [%s] in key store %s.",
                                KeyStore.getDefaultType(), certStore);
                        throw new IllegalStateException(errorMessage, e);
                    } finally {
                        Consumer<IOException> emptyExceptionHandler = ex -> {
                        };
                        IOUtils.closeQuietly(inputStream, emptyExceptionHandler);
                    }
                }, EXECUTOR);
    }

```

If this method wasn't using the asynchronous completable future APIs the initial bug - deadlock - would have not happened.

In conclusion, I don't think this is a good optimization because its not general and the implementation is unnecessarily complicated. If you still want to have this optimization, then you can have a cleaner implementation with a Guava loading cache that will expire the cache every 3 seconds, that will eliminate most of the async  code.

Reply:
The patch tries to achieve two things: Cache a file and read the file asynchronously. I think you still pursue the caching part, but without the asynchronous code because of the aforementioned reasoning. Notice that not only can the file be cached, but you can also cache the File.exists check as well.